src/dragonball/src/api/v1/vmm_action.rs

```diff
@@ -634,3 +634,626 @@ fn handle_cpu_topology(
+    }
+
+    #[should_panic]
+    #[test]
+    fn test_vmm_action_disconnected() {
+        let (to_vmm, from_api) = channel();
+        let (to_api, _from_vmm) = channel();
+        let vmm = Arc::new(Mutex::new(create_vmm_instance()));
+        let mut vservice = VmmService::new(from_api, to_api);
+        let epoll_mgr = EpollManager::default();
+        let mut event_mgr = EventManager::new(&vmm, epoll_mgr).unwrap();
+        let mut v = vmm.lock().unwrap();
+
+        drop(to_vmm);
+        vservice.run_vmm_action(&mut v, &mut event_mgr).unwrap();
+    }
+
+    #[test]
+    fn test_vmm_action_config_boot_source() {
+        skip_if_not_root!();
+
+        // invalid state
+        let req = VmmAction::ConfigureBootSource(BootSourceConfig::default());
+        check_request(req, InstanceState::Running, |result| {
+            if let Err(VmmActionError::BootSource(
+                BootSourceConfigError::UpdateNotAllowedPostBoot,
+            )) = result
+            {
+                let err_string = format!("{}", result.unwrap_err());
+                let expected_err = String::from(
+                    "failed to configure boot source for VM: \
+                    the update operation is not allowed after boot",
+                );
+                assert_eq!(err_string, expected_err);
+            } else {
+                panic!();
+            }
+        });
+
+        // invalid kernel file path
+        let req = VmmAction::ConfigureBootSource(BootSourceConfig::default());
+        check_request(req, InstanceState::Uninitialized, |result| {
+            if let Err(VmmActionError::BootSource(BootSourceConfigError::InvalidKernelPath(_))) =
+                result
+            {
+                let err_string = format!("{}", result.unwrap_err());
+                let expected_err = String::from(
+                    "failed to configure boot source for VM: \
+                    the kernel file cannot be opened due to invalid kernel path or invalid permissions: \
+                    No such file or directory (os error 2)");
+                assert_eq!(err_string, expected_err);
+            } else {
+                panic!();
+            }
+        });
+
+        //success
+        let mut boot_source_config = BootSourceConfig::default();
+        let kernel_file = TempFile::new().unwrap();
+        boot_source_config.kernel_path = kernel_file.as_path().to_str().unwrap().to_string();
+        let req = VmmAction::ConfigureBootSource(boot_source_config);
+        check_request(req, InstanceState::Uninitialized, |result| {
+            assert!(result.is_ok());
+        });
+    }
+
+    #[test]
+    fn test_vmm_action_set_vm_configuration() {
+        skip_if_not_root!();
+
+        // invalid state
+        let req = VmmAction::SetVmConfiguration(VmConfigInfo::default());
+        check_request(req, InstanceState::Running, |result| {
+            if let Err(VmmActionError::MachineConfig(VmConfigError::UpdateNotAllowedPostBoot)) =
```

Comment:
Is the `if let Err(xxx) = result` needed? can this be simplified by:

```diff
-             if let Err(VmmActionError::MachineConfig(VmConfigError::UpdateNotAllowedPostBoot)) =
-                 result
-             {
-                 let err_string = format!("{}", result.unwrap_err());
-                 let expected_err = String::from(
-                     "failed to set configuration for the VM: \
-                     update operation is not allowed after boot",
-                 );
-                 assert_eq!(err_string, expected_err);
-             } else {
-                 panic!("unexpect");
-             }
---
+             assert!(result.is_err());
+             let err_string = format!("{}", result.unwrap_err());
+             let expected_err = String::from(
+                 "failed to set configuration for the VM: \
+                 update operation is not allowed after boot",
+             );
+             assert_eq!(err_string, expected_err);

```

Reply:
fixed.